tests/unit/legate.core/test_machine.py

```diff
@@ -0,0 +1,305 @@
+        r3 = r1 & r2
+        assert r3 == ProcessorRange.create(ProcessorKind.GPU, 1, 1, 0)
+        assert len(r3) == 0
+
+    def test_intersection_invalid(self) -> None:
+        r1 = ProcessorRange.create(ProcessorKind.CPU, 1, 1, 0)
+        r2 = ProcessorRange.create(ProcessorKind.GPU, 1, 1, 0)
+        err_msg = (
+            "Intersection between different processor kinds: " "CPU and GPU"
+        )
+        with pytest.raises(ValueError, match=err_msg):
+            r1 & r2
+
+    def test_empty_slice(self) -> None:
+        r = ProcessorRange.create(ProcessorKind.GPU, 1, 2, 5)
+        assert len(r.slice(slice(0, 0))) == 0
+        assert len(r.slice(slice(None, 0))) == 0
+        assert len(r.slice(slice(4, None))) == 0
+
+        r = ProcessorRange.create(ProcessorKind.GPU, 1, 3, 1)
+        assert len(r.slice(slice(0, 0))) == 0
+        assert len(r.slice(slice(None, 0))) == 0
+        assert len(r.slice(slice(4, None))) == 0
+
+    def test_nonempty_slice(self) -> None:
+        r = ProcessorRange.create(ProcessorKind.GPU, 1, 2, 5)
+        assert len(r.slice(slice(None))) == len(r)
+        for i in range(len(r)):
+            assert len(r.slice(slice(i))) == i
+            assert len(r.slice(slice(i, None))) == len(r) - i
+
+        r = ProcessorRange.create(ProcessorKind.GPU, 1, 3, 1)
+        assert len(r.slice(slice(None))) == 0
+        for i in range(len(r)):
+            assert len(r.slice(slice(i))) == 0
+            assert len(r.slice(slice(i, None))) == 0
+
+    def test_invalid_slice(self) -> None:
+        r = ProcessorRange.create(ProcessorKind.GPU, 1, 2, 4)
+        with pytest.raises(ValueError, match="The slicing step must be 1"):
+            r.slice(slice(None, None, 2))
+
+    def test_pack(self) -> None:
+        r = ProcessorRange(ProcessorKind.CPU, 3, 4, 5)
+        buf = BufferBuilder()
+        r.pack(buf)
+        packed = buf.get_string()
+        assert packed is not None
+        assert buf.get_size() == 12
+        v1, v2, v3 = struct.unpack("III", packed)
+        assert v1 == 3
+        assert v2 == 4
+        assert v3 == 5
+
+
+RANGES = [
+    ProcessorRange.create(ProcessorKind.CPU, 4, 1, 3),
+    ProcessorRange.create(ProcessorKind.OMP, 2, 0, 3),
+    ProcessorRange.create(ProcessorKind.GPU, 3, 3, 6),
+]
+EMPTY_RANGE = ProcessorRange.create(ProcessorKind.GPU, 1, 1, 0)
+
+
+class TestMachine:
+    def test_empty_machine(self) -> None:
+        m = Machine([])
+        assert m.preferred_kind == ProcessorKind.CPU
+        assert len(m) == 0
+        assert len(m.get_processor_range(ProcessorKind.CPU)) == 0
+        err_msg = "Illegal to get a node range of an empty processor range"
+        with pytest.raises(ValueError, match=err_msg):
+            m.get_node_range()
+        with pytest.raises(ValueError, match=err_msg):
+            m.get_node_range(ProcessorKind.GPU)
+
+    def test_eq(self) -> None:
+        m1 = Machine(RANGES[:-1])
```

Comment:
This indexing seems a bit delicate in case `RANGES` ever changes. Suggest defining individual vars e.g. `CPU`, `OMP`, `GPU` then be explicit:
```
m1 = Machine([CPU, OMP])
m2 = Machine([CPU, OMP])
assert m1 == m2
```

(or `CPU_R` or `CPU_RANGE` etc  to even more explicit)

Reply:
https://github.com/nv-legate/legate.core/pull/457/commits/5739b8a273d137921727be956abc2b2e0515662c